pandas/core/categorical.py

```diff
@@ -1665,16 +1669,28 @@ def fillna(self, value=None, method=None, limit=None):
 
         else:
 
-            if not isna(value) and value not in self.categories:
-                raise ValueError("fill value must be in categories")
+            # If value is a dict or a Series (a dict value has already
+            # been converted to a Series)
+            if isinstance(value, ABCSeries):
+                if not value[~value.isin(self.categories)].isna().all():
+                    raise ValueError("fill value must be in categories")
 
-            mask = values == -1
-            if mask.any():
-                values = values.copy()
-                if isna(value):
-                    values[mask] = -1
-                else:
-                    values[mask] = self.categories.get_loc(value)
+                values_codes = _get_codes_for_values(value, self.categories)
+                indexer = np.where(values_codes != -1)
+                values[indexer] = values_codes[values_codes != -1]
+
+            # If value is not a dict or Series it should be a scalar
+            else:
+                if not isna(value) and value not in self.categories:
```

Comment:
you need to test if its a ``is_scalar`` *before* you use ``isna``, because if its not then this will raise a different error (add a test for that as well). (you can make this an elif which ``is_scalar`` is True and raise in an ``else``

Reply:
Thanks, that's a good idea.